content/architecture/nanoclr-stub-args.md

```diff
@@ -0,0 +1,209 @@
+
+Those `NANOCLR_CHECK_HRESULT` and `NANOCLR_EXIT_ON_SUCCESS` macros allow you to check if a function or an expressing is a failure or success and then as we've seen previously go to `nanoCLR_Cleanup`. This is used a lot when you are calling other similar function returning as well an `HRESULT`.
+
+```cpp
+#define NANOCLR_CHECK_HRESULT(expr)					{ if(FAILED(hr = (expr))) NANOCLR_LEAVE(); }
+#define NANOCLR_EXIT_ON_SUCCESS(expr)				{ if(SUCCEEDED(hr = (expr))) NANOCLR_LEAVE(); }
+#define NANOCLR_SET_AND_LEAVE(expr)					{ hr = (expr); NANOCLR_LEAVE(); }
+```
+
+The `NANOCLR_SET_AND_LEAVE` function will just set the `HRESULT` and go to `nanoCLR_Cleanup`.
+
+You will find a detailed list of typical HRESULT in the `src\CLR\Include\nf_errors_exceptions.h` file. FAILED and SUCCEEDED are defined like this:
+
+```cpp
+#define SUCCEEDED(Status) ((HRESULT)(Status) >= 0)
+#define FAILED(Status) ((HRESULT)(Status)<0)
+```
+
+### NANOCLR_MSG_SET_AND_LEAVE and NANOCLR_MSG1_SET_AND_LEAVE
+
+Those 2 macros are defined like this:
+
+```cpp
+#define NANOCLR_MSG_SET_AND_LEAVE(expr, msg)		{ hr = (expr); NANOCLR_LEAVE(); }
+#define NANOCLR_MSG1_SET_AND_LEAVE(expr, msg, arg)	{ hr = (expr); NANOCLR_LEAVE(); }
+```
+
+So they allow you to setup the return element and leave. Those are so far used only in the `src\CLR\Core\TypeSystem.cpp` file. That said, nothing prevent you to use them as well.
+
+## CLR_RT_StackFrame &stack
+
+At every call to one of those native functions, the stack is passed thru a structure called CLR_RT_StackFrame. The definition can be found in `src\CLR\Include\nanoCLR_Runtime.h`. It would be too long to go thru all the functions and properties. In this description we will focus only on few elements.
+
+### If your function is in a static class
+
+In, this case you **won't** have the class stack available. Stack class is only available for non static class.
+
+### Getting and checking the stack in a non static class
+
+The pattern to use is the following:
+
+```cpp
+CLR_RT_HeapBlock* pThis = stack.This();
+FAULT_ON_NULL(pThis);
+```
+
+`FAULT_ON_NULL` is a macro similar to the onces presented in the previous section which will check of the stack is null or not. In case of null it will go to the `nanoCLR_Cleanup` label and set the `HRESULT` to a null fault.
+
+The definition can be found in `src\CLR\Include\nanoCLR_Interop.h`.
+
+```cpp
+#define FAULT_ON_NULL(ptr)							if(!(ptr)) NANOCLR_SET_AND_LEAVE(CLR_E_NULL_REFERENCE)
+#define FAULT_ON_NULL_ARG(ptr)						if(!(ptr)) NANOCLR_SET_AND_LEAVE(CLR_E_ARGUMENT_NULL)
+```
+
+You can use those macro for arguments too. We will see this in one of the following section.
+
+### Getting any exposed field from the stack
+
+Once the stack is valid, you can get any exposed field. Here is a typical example:
+
+```cpp
+int pinNumber = (int)(pThis[Library_sys_dev_pwm_native_System_Device_Pwm_PwmChannel::FIELD___pinNumber].NumericByRef().u4);
+```
+
```

Comment:
In case you're accessing one of the fields of the "current" class, there is no need to use the fully qualified name with the namespace. It would be OK to just use `pThis[FIELD___pinNumber].NumericByRef().u4`.

Reply:
Yes, that I know, I just wrote it's recommended to use the long name to avoid any issue. Now, I can surely adjust and make it more clear.

Reply:
No it's OK. I can see that you have a detailed explanation about this namespace in the next paragraph.